src/common/FastRational.h

```diff
@@ -840,12 +940,29 @@ inline void substractionAssign(FastRational& a, const FastRational& b) {
 
 inline void multiplicationAssign(FastRational& a, const FastRational& b) {
     if (a.wordPartValid() && b.wordPartValid()) {
-        lword common1 = gcd(absVal(a.num), b.den);
-        lword common2 = gcd(a.den, absVal(b.num));
+        uword common1 = gcd(absVal(a.num), b.den);
+        uword common2 = gcd(a.den, absVal(b.num));
         word zn;
         uword zd;
-        CHECK_WORD(zn, lword(a.num/common1) * (b.num/common2));
-        CHECK_UWORD(zd, ulword(a.den/common2) * (b.den/common1));
+        // Without the absVal, this fails for a.num < 0 when common1 > 1 and b.num < 0 when common2 > 1 since the result of the division is unsigned.
+        CHECK_WORD(zn, lword(common1 > 1 ? absVal(a.num)/common1 : absVal(a.num)) * (common2 > 1 ? absVal(b.num)/common2 : absVal(b.num)));
+        CHECK_UWORD(zd, ulword(common2 > 1 ? a.den/common2 : a.den) * (common1 > 1 ? b.den/common1 : b.den));
+
+        bool b_num_lt_0 = b.num < 0;
+        bool a_num_ge_0 = a.num >= 0;
+        bool b_num_gt_0 = b.num > 0;
+        bool a_num_le_0 = a.num <= 0;
+
+        if ((b_num_lt_0 && a_num_ge_0) || (b_num_gt_0 && a_num_le_0)) { //CHECK_WORD(zn, (lword)(-absVal(zn)));
+            lword tmp = -(lword)(absVal(zn));
```

Comment:
Again here, the code above seems to guarantee now that `zn` will be positive, since `absVal` is used in its computation, so here it seems redundant.

Reply:
I agree.